contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -192,96 +225,100 @@ type EphemeralContainer struct {
         ...
-        EphemeralContainerStatuses []v1.ContainerStatus
+        // List of user-initiated ephemeral containers that have been run in this pod.
+        // +optional
+        EphemeralContainers []EphemeralContainer `json:"commands,omitempty" protobuf:"bytes,11,rep,name=ephemeralContainers"`
+
 }
 ```
 
-This status is only populated for Debug Containers, but there's interest in
-tracking status for traditional exec in a similar manner.
-
-Note that `Command` and `Args` would have to be tracked in the status object
-because there is no spec for Debug Containers or exec. These must either be made
-available by the runtime or tracked by the kubelet. For Debug Containers this
-could be stored as runtime labels, but the kubelet currently has no method of
-storing state across restarts for exec. Solving this problem for exec is out of
-scope for Debug Containers, but we will look for a solution as we implement this
-feature.
-
-`EphemeralContainerStatuses` is populated by the kubelet in the same way as
-regular and init container statuses. This is sent to the API server and
-displayed by `kubectl describe pod`.
+The kubelet should be able to construct a complete `PodStatus` with no prior
+state using information stored in the container runtime.
+`EphemeralContainer.Status` introduces no new data, but the kubelet must also
+now populate `EphemeralContainer.Spec` &
+`EphemeralContainer.TargetContainerName`.
+
+The kubelet already persists container metadata as CRI
+[labels](https://github.com/kubernetes/kubernetes/blob/v1.10.0-alpha.0/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L606)
+and
+[annotations](https://github.com/kubernetes/kubernetes/blob/v1.10.0-alpha.0/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L613).
+The entire v1.Container used in the request will be serialized and stored as a
+runtime annotation. The value of `TargetContainerName` will be stored as a
+runtime label. Persisting this data in the runtime means it will survive kubelet
+restarts.
+
+At least for the Docker runtime, this is [an intended use of docker
+labels](https://docs.docker.com/engine/userguide/labels-custom-metadata/#value-guidelines).
+Docker does not document the maximum length of labels in its API. Empirically,
+it supports up to the 64K constraint of the docker client's `bufio.Scanner`
+size. Because the container spec may be examined in security sensitive contexts
+like admission control, we will conservatively limit the size of the spec to 32K
+and add a 32K minimum label length test to runtime qualification.
+
+`EphemeralContainer` is populated by the kubelet in the same way as regular
+container statuses. This is sent to the API server and displayed by `kubectl
+describe pod`.
 
 ### Creating Debug Containers
 
-1.  `kubectl` invokes the exec API as described in the preceding section.
+1.  `kubectl` invokes the new API as described in the preceding section.
 1.  The API server checks for name collisions with existing containers, performs
```

Comment:
I'm assuming the kubelet also checks for name collisions with existing containers or existing debug containers, and that the API server validation prevents name collisions when the kubelet updates pod status with ephemeral container status

Reply:
That's correct. Checking for collisions in the api server provides a quick, user-friendly error while the kubelet resolves any races in its dedicated pod worker.

I've updated this section to clarify.